### ISSUE TITLE
Doesn't work for all strings

### ISSUE THREAD

**Issue by jeremycw** [anismiles/websocket-android-phonegap#2]
The websocket fails to receive the following string:

{"script":"// 49\n// 32\n\n//stopTimer();\n\nif(!mobileDevice(navigator.userAgent, true)){\n $('#status_bar_container').html('\n\n<div id=\\\"status_bar\\\">\n \n <script type=\\\"text/javascript\\\" charset=\\\"utf-8\\\">\n \n <\/script>\n \n\t<table id=\\\"status_table\\\" cellpadding=\\\"0\\\" cellspacing=\\\"0\\\">\n\t\t<tr>\n\t\t\t\n\t\t\t<td id=\\\"health_bar\\\" title=\\\"Health Points: Go to the Medic.\\\" onclick=\\\"goToUrl(\\'http://fb.tunnel.uken.com:3040/regen_chamber?class=loader\\')\\\">\n <div class=\\\"txt_box\\\" onmouseover=\\\"$(\\'#HP_time\\').hide();$(\\'#HP_link\\').css(\\'display\\', \\'inline-block\\');\\\" onmouseout=\\\"$(\\'#HP_link\\').hide();$(\\'#HP_time\\').css(\\'display\\', \\'inline-block\\');\\\">\n\t<span id=\\\"HP_amount\\\" class=\\\"status_numbers\\\">\n\t\t84 | 100\t<\/span><br />\n\t<span id=\\\"HP_time\\\" style=>\n \n\t +10 in 3:58\n \n <\/span>\n <span id=\\\"HP_link\\\" style=\\\"display: none;\\\" >\n <a href=\\\"/regen_chamber\\\" class=\\\"loader\\\">heal<\/a> \n <\/span>\n<\/div>\n\t\t\t<\/td>\n\t\t\t<td id=\\\"money_bar\\\" title=\\\"Money: Go to the Bank.\\\" onclick=\\\"goToUrl(\\'http://fb.tunnel.uken.com:3040/vault?class=loader\\')\\\">\n <div class=\\\"txt_box\\\" onmouseover=\\\"$(\\'#money_time\\').hide();$(\\'#money_link\\').css(\\'display\\', \\'inline-block\\');\\\" onmouseout=\\\"$(\\'#money_link\\').hide();$(\\'#money_time\\').css(\\'display\\', \\'inline-block\\');\\\">\n\t<span id=\\\"money\\\" class=\\\"money status_numbers\\\" value=\\\"10739\\\">\n\t\t<span class=\\'price\\'>$10,739<\/span>\n\t<\/span><br />\n\t\n\t\t<span id=\\\"money_time\\\" class=\\\"help\\\" title=\\\"Hourly Cashflow: $0\\\">\n\t\t more in 49:15\n <\/span>\n <span id=\\\"money_link\\\" style=\\\"display: none;\\\">\n <a href=\\\"http://fb.tunnel.uken.com:3040/vault\\\" class=\\\"loader\\\">go to bank<\/a> \n <\/span>\n\t\n<\/div>\n\t\t\t<\/td>\n\t\t\t<td id=\\\"energy_bar\\\" title=\\\"Energy: needed for Missions.\\\" onclick=\\\"goToUrl(\\'http://fb.tunnel.uken.com:3040/missions?class=loader\\')\\\">\n <div class=\\\"txt_box\\\">\n\t<span id=\\\"Energy_amount\\\" class=\\\"status_numbers\\\">\n 10 | 10 <\/span><br />\n <span id=\\\"Energy_time\\\">\n \n Energy full\n \n <\/span>\n<\/div>\n\t\t\t<\/td>\n\t\t\t<td id=\\\"stamina_bar\\\" title=\\\"Stamina: needed for Battles.\\\" onclick=\\\"goToUrl(\\'http://fb.tunnel.uken.com:3040/battle?class=loader\\')\\\">\n <div class=\\\"txt_box\\\">\n\t<span id=\\\"Stamina_amount\\\" class=\\\"status_numbers\\\">\n 4 | 4 <\/span><br />\n <span id=\\\"Stamina_time\\\">\n \n Ammo full\n \n <\/span>\n<\/div>\n\t\t\t<\/td>\n\t\t\t<td id=\\\"science_bar\\\" title=\\\"Go to Black Market.\\\" onclick=\\\"goToUrl(\\'http://fb.tunnel.uken.com:3040/lab?class=loader\\')\\\">\n\t\t\t\t<div class=\\\"txt_box\\\">\n\t\t\t\t\t<span class=\\\"status_numbers\\\"><span id=\\\"science_points\\\">22<\/span> gold bars<\/span>\n\t <br/>\n\t <span><a href=\\\"http://fb.tunnel.uken.com:3040/lab\\\" class=\\\"loader\\\">go to black market<\/a><\/span>\n\t\t\t\t<\/div>\n\t\t\t<\/td>\n\t\t\t<td id=\\\"xp_bar\\\">\n \n<div id=\\\"xp_wrapper\\\">\n <div class=\\\"xp_txt\\\">\n \t<span class=\\\"status_numbers\\\">Level: 13<\/span>\n\n \t\n \t\t<a href=\\\"/profile\\\" class=\\\"loader\\\">upgrade!<\/a><\/span>\n \t\n\n \t<div title=\\\"Next Level: 31 xp\\\"><\/div>\n <\/div>\n\n <div id=\\\"progress\\\">\n <div style=\\\"width: 35.4166666666667%; height:inherit\\\"><\ V/websocket( 6399): Received a message: {"script":"$(\"#attack_alert_collection\").append( '<script type=\\\"text/javascript\\\">\n $(document).ready(function() {\n UkenLang.populate();\n });\n<\/script>\n\n\n\n \n\n\n<div style=\\\"display:none\\\" id=\\\"alert_116\\\" class=\\\"attacker_alert alert\\\">\n <p class=\\\"alert_title\\\">ATTACKED<\/p>\n\t\n\tYou Were Just Attacked by <a href=\\\"/superhero/32?class=loader\\\">poop<\/a> and <span style=\\\"color:#08FF00\\\">WON<\/span>!!\n<\/div>\n\n' );\n\nif(mobileDevice(navigator.userAgent, true))\n $('#alert_116 .alert_title').hide();\n\n$('#alert_116').slideDown().effect(\"highlight\", {}, 1000);\n\nvar a = $('.attacker_alert').length;\n\nif (a > 3) {\n\tvar a = $(\".attacker_alert:first-child\");\n\t\n\ta.slideUp('slow', function() {\n\t\ta.remove();\n\t});\n} else if(window.ContentScroller) {\n //support for scroller/fixed positioning\n window.setTimeout(function(){window.ContentScroller.updateViewport()}, 1050);\n}\nUkenLang.populate();\n\n// if (a == 1) {\n\t// $('#alerts').after('<a href=\\\"#\\\" id=\\\"clear_alerts\\\" class=\\\"no-top\\\">clear out</a>')\n// }\n\n// $('#clear_alerts').click(function() {\n\t// $('.attacker_alert').remove();\n\t// $('#clear_alerts').remove();\n// });\n"}

**Comment by anismiles** [comment]
Can you share some pieces of your code? The above ones are too messy to make any sense out of this.

**Comment by jeremycw** [comment]
Sorry, I probably didn't explain myself well enough. That's not my websocket code, that's the string I'm trying to send through the socket.

say I did this:

var ws = new WebSocket("ws://someserver.com:8003");
var someNastyStringWithJsonAndHtml = '{"script":"// 49n// 32nn//stopTimer();nnif(!mobileDevice(navigator.userAgen... }';
ws.send(someNastyStringWithJsonAndHtml);

onmessage will never fire for any connected clients.

in your code here:

private String buildJavaScriptData(String event, String msg) {
    String _d = "javascript:WebSocket." + event + "(" + "{" + "\"_target\":\"" + id + "\","
            + "\"_data\":'" + msg + "'" + "}" + ")";
    return _d;
}

the msg is creating unparsable JSON so that it's erroring out on the JavaScript side.

I forked your project and solved the issue by base64 encoding the message on the native side before passing it to the JS through loadUrl and then decoding it in JS. Not the most elegant solution, but it works.

A better solution would probably be to add another javascript interface onto webkit wich get's the message. In the native code store the message after it's received and expose a getMsg() interface to the javascript and then instead of passing in the message through loadUrl, just use loadUrl to trigger the javascript event and let the event call getMsg()
